utils/rate_limiter.go

```diff
@@ -47,18 +49,42 @@ type rateLimiter struct {
 //
 // It can also be used to limit the rate of traffic in bytes, by setting creditsPerSecond to desired throughput
 // as bytes/second, and calling CheckCredit() with the actual message size.
-func NewRateLimiter(creditsPerSecond, maxBalance float64) RateLimiter {
-	return &rateLimiter{
+func NewRateLimiter(creditsPerSecond, maxBalance float64) *RateLimiterImpl {
+	return &RateLimiterImpl{
 		creditsPerSecond: creditsPerSecond,
-		balance:          maxBalance,
+		balance:          maxBalance * rand.Float64(),
 		maxBalance:       maxBalance,
 		lastTick:         time.Now(),
-		timeNow:          time.Now}
+		timeNow:          time.Now,
+	}
+}
+
+// Update updates the rate limiter with new configurations. This allows the balance to carry over as the rate changes.
+func (b *RateLimiterImpl) Update(creditsPerSecond, maxBalance float64) {
```

Comment:
Had to make the struct public so that this Update function can be used in the other directory. I didn't want to add the function to the interface as I didn't feel like it belonged there.

Reply:
I would suggest a different approach - move the implementation to `internal` package. Then you can continue calling the struct RateLimiter (I swear this is the first time I see SomethingImpl in Go!).

Also, I am not sure why we need the `util.RateLimiter` interface - do we have some public funcs that accept it?  I thought the samplers always instantiate it internally, meaning that the interface can also be private to the sampler.